### PR TITLE
AssociatedTypeInference: Initial refactoring of abstract type witness inference 

### PR DESCRIPTION
**Pull request by AnthonyLatsis**

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please clean smoke test macOS

File: lib/Sema/TypeCheckProtocolInference.cpp [hunk 1]
```diff
@@ -2091,6 +2091,245 @@ auto AssociatedTypeInference::solve(ConformanceChecker &checker)
...
+      if (this->TypeWitnesses.count(depTy->getName())) {
+        return addEquivalence(name, depTy->getName());
+      }
+    } else {
+      while (depTy->getBase()->is<DependentMemberType>()) {
+        depTy = depTy->getBase()->castTo<DependentMemberType>();
+      }
+
+      // Equivalences of the form 'Self.X == Self.X.*' do not contribute
+      // to solving the system, so just ignore them.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocolInference.cpp:2216]: What about Self.X.Y == Self.X.Y.*, etc?

File: lib/Sema/TypeCheckProtocolInference.cpp [hunk 2]
```diff
@@ -922,6 +922,75 @@ AssociatedTypeInference::computeAbstractTypeWitness(
...
+  for (auto *const conformedProto :
+       adoptee->getAnyNominal()->getAllProtocols()) {
+    // FIXME: We could look through the requirement signatures of other
+    // protocols as well.
+    if (conformedProto != proto && !conformedProto->inheritsFrom(proto))
+      continue;
+
+    // FIXME: The RequirementMachine will assert on re-entrant construction.
+    // We should find a more principled way of breaking this cycle.
+    if (ctx.isRecursivelyConstructingRequirementMachine(
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocolInference.cpp:853]: There's a merge conflict here.

File: test/decl/protocol/req/associated_type_inference_fixed_type.swift [hunk 3]
```diff
@@ -435,19 +435,199 @@ do {
...
+protocol Q34b: P34b, P34a {}
+protocol Q34c: P34a, P34b {
+  associatedtype A // expected-note {{protocol requires nested type 'A'; do you want to add it?}}
+}
+do {
+  // FIXME: should really be ambiguous (source-breaking)?
+  // CHECK-LABEL: Abstract type witness system for conformance of Conformer1 to P34a: {
+  // CHECK-NEXT: A => Void,
+  // CHECK-NEXT: }
+  struct Conformer1: Q34a {}
```
Comment (slavapestov) [inline test/decl/protocol/req/associated_type_inference_fixed_type.swift:621]: It looks like this is sensitive to the order of protocols in the inheritance clause, which might be weird. Perhaps you should canonicalize the list of inherited protocols when walking (or walk over conformance requirements on 'Self' in the requirement signature instead, which are in canonical order).

**Review by slavapestov — APPROVED** [review]
Does this PR fix the problem where associated types must be re-stated in inherited protocols?

Eg, this doesn't type check today:

```
protocol Base {
  associatedtype A
}

protocol Derived : Base {
  func foo() -> A
}

struct S : Derived {
  func foo() -> Int { return 0 }
}
```

You need to pointlessly re-declare the 'associatedtype A' inside 'protocol Derived'. It would be nice to make this work.

**Comment by slavapestov** [comment]
@AnthonyLatsis can you run source compat testing on this once you resolve the merge conflict?

File: lib/Sema/TypeCheckProtocolInference.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckProtocolInference.cpp:2216]: Anything other than `Self.X := Y` is not considered at all.

**Comment by AnthonyLatsis** [comment]
> Does this PR fix the problem where associated types must be re-stated in inherited protocols?

I am afraid not, I haven't touched inference via value witnesses yet.

> @AnthonyLatsis can you run source compat testing on this once you resolve the merge conflict?

I sure will!

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@slavapestov source compatibility failures are unrelated, but in isolation this is in fact a source breaking change. There is a group of conformances we are not yet prepared to deal with for which associated type inference now fails. A selected few from these compile today:

```swift
protocol P { associatedtype A = Int }
protocol Q { associatedtype B }
protocol R: P, Q where A == B {}

struct S: R {} // A := Int, B := Int
```
And an even smaller fraction actually resolve to the correct type witnesses:

```swift
protocol P { associatedtype B = Int }
protocol Q { associatedtype A }
protocol R: P, Q where A == B {}

struct S<A>: R {} // A := A (generic parameter), B := A
```

They all have two things in common: you need more information than just the same-type constraints to decide on a type witness, and the type witnesses of one conformance cannot be resolved independently of another in general.
I reckon that having something fail before it starts working properly is better than a sudden change in behavior. Still, I would like to hear your thoughts on whether we should gate this behind a flag.

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
The probability of someone relying on the source entailing compatibility nuance I mentioned is really low, given that these cases just started "working" in the last minor release or so, so I'm going to merge this and try do deal with the existing bigger issue in 5.7. Then again, we can always defer or revert this if something goes left.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci clean smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci clean smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Pull request merged**